src/utils/fluid_sys.c

```diff
@@ -386,11 +386,19 @@ unsigned int fluid_curtime(void)
 double
 fluid_utime (void)
 {
+#if defined(WIN32)      /* Window using intel performance counters  */
+	LARGE_INTEGER Freq;	/* Performance Frequency */
+	LARGE_INTEGER PerfCpt;
+	QueryPerformanceFrequency(&Freq);  /* Frequency value */
+	QueryPerformanceCounter(&PerfCpt); /* Counter value */
+	return PerfCpt.QuadPart * 1000000.0/Freq.QuadPart; /* time in micros */
+#else
```

Comment:
I think we should get rid of the special treatment for win32 by switching to g_get_monotonic_time(), which should provide the required resolution: https://github.com/GNOME/glib/blob/1e6803be3b3ec5005335f1fa7d53e57bbc0b8bba/glib/gmain.c#L2744

Reply:
> switching to g_get_monotonic_time(), which should provide the required resolution:

I have tried this to verify if the result are the same as actual `fluid_utime() (L392 to L394)`.
Unfortunately the value returned by  `g_get_monotonic_time()` is not the same and the resolution is far less than `fluid_utime()` .  Some results follow:
```
now_fluid_utime():                        16055873753.508 micros
now_g_get_monotonic_time():          1520527285078125.000 micros 
```
```
// Duration measured using fluid_utime():
delta_fluid_utime = now_fluid_utime - prev_fluid_utime

// Duration measured using get_monotonic_time():
delta_g_get_monotonic_time = now_g_get_monotonic_time - g_get_monotonic_time

// Difference between these durations:
diff_glib_fluid_utime = delta_fluid_utime - delta_g_get_monotonic_time

delta_fluid_utime:                            1001003.479 micros
delta_g_get_monotonic_time:                   1000000.000 micros
diff_glib_fluid_utime:                           1003.479 micros  <-  error 1 ms 
```
`fluid_utime()` return the result in micros as a real keeping the resolution of the hardware counter. For example with a frequency of  3579545 Hz , the resolution is  0.279 micros.
Using `get_monotonic_time()` we get often an error about 1 ms between duration of 1 second. 
The value returned by `get_monotonic_time()` is not the value of the actual hardware counter. This function doesn't behave as we expect.

Reply:
But if you measure for 1s and `delta_g_get_monotonic_time` gives you the perfect second, why do you think it's not exact? I believe that `fluid_utime()` is not exact then. Perhaps because retrieving `QueryPerformanceFrequency(&Freq);` every time is expensive? Also microsofts doc recommend to cache `Freq`, so glibs  `g_get_monotonic_time()` seems like the correct implementation.

Reply:
I must admit that previous example is not clear and don't reveals what happens with 'g_get_monotonic_time()'.

I investigated further calling actual `fluid_utime()`, `g_get_monotonic_time()` and `g_get_current_time()` at the same time:

```
GTimeVal gtimeval;
double gctime;
double now_fluid_utime_cpt = fluid_utime();
double now_glib_cpt = (double) g_get_monotonic_time();
g_get_current_time(&gtimeval);

printf("fluid_utime():              %30.3f micros\n",now_fluid_utime_cpt);
printf("g_get_monotonic_time():     %30.3f micros\n",now_glib_cpt);
gctime = (double)gtimeval.tv_sec * 1000000.0 + (double)gtimeval.tv_usec;
printf("g_get_current_time():       %30.3f micros \n",gctime);
```
The result of print is:
```	
fluid_utime():                             23961936852.868 micros
g_get_monotonic_time():               1520640580187500.000 micros
g_get_current_time():                 1520640580187500.000 micros
```	
`g_get_monotonic_time()` returns the same result that `g_get_current_time()` and not the  same as `fluid_utime()` (that uses  `QueryPerformanceCounter()` ). So it seems that `g_get_monotonic_time()` make use of the same system call than `g_get_current_time()`.

`g_get_current_time()` make use of `GetSystemTimeAsFileTime()` that obtains the current system date and time that is less precise than `QueryPerformanceCounter()`.
***
>Perhaps because retrieving QueryPerformanceFrequency(&Freq); every time is expensive? Also microsofts doc recommend to cache Freq.

Yes, it a good practice to cache Freq even if `QueryPerformanceFrequency()` isn't too expensive (1.509 micros).

The  duration results show that caching is a good thing:
- fluid_utime() without caching: 2.982 micros.
- fluid_utime() with caching: 1.550 micros.
- QueryPerformanceFrequency(): 1.523 micros.

Reply:
> g_get_monotonic_time() returns the same result that g_get_current_time()

I kept looking at the implementation on github and just couldnt believe it. But you're right. And now I have the answer: You need at least glib 2.53.3 to get accurate results. Here's the bug report: https://bugzilla.gnome.org/show_bug.cgi?id=783340

So, what shall we do? I still believe reprogramming `g_get_monotonic_time()` is not a good way to go. Bumping the minimum required glib version to 2.53.3 would be an overkill. And after all, this topic is only relevant for us developers who want to profile from time to time. How about the following solution:

```c
double
fluid_utime (void)
{
#if GLIB_MAJOR_VERSION == 2 && GLIB_MINOR_VERSION >= 28
  /* use high precision monotonic clock if available and warn
   * if this clock is actually implemented as low prec. clock,
   * i.e. in case glib is too old,
   * see: https://bugzilla.gnome.org/show_bug.cgi?id=783340
   */
#if defined(WITH_PROFILING) &&\
    defined(WIN32) &&\
    GLIB_MINOR_VERSION <= 53 &&\
    GLIB_MICRO_VERSION < 3
#warning "Time measurements may be inaccurate. Get at least glib 2.53.3 for high precision monotonic time."
#endif
    
  return g_get_monotonic_time();
  
#else
  // fallback to less precise clock
  GTimeVal timeval;
  g_get_current_time (&timeval);

  return (timeval.tv_sec * 1000000.0 + timeval.tv_usec);
#endif
}
```

Reply:
How about this solution for which `glib have precedence` unless the used version is unable to do the job.
```
double
fluid_utime (void)
{
#if GLIB_MAJOR_VERSION == 2 && GLIB_MINOR_VERSION >= 28
/* use high precision monotonic clock if available and warn
   * if this clock is actually implemented as low prec. clock,
   * i.e. in case glib is too old,
   * see: https://bugzilla.gnome.org/show_bug.cgi?id=783340
   */
#if defined(WITH_PROFILING) &&  defined(WIN32) &&\
	/* glib < 2.53.3 */\
    (GLIB_MINOR_VERSION < 53 || (GLIB_MINOR_VERSION == 53 &&  GLIB_MICRO_VERSION < 3))
#pragma message( "Warning: fluid_utime()" )
#pragma message( "This glib version is unable to use high precision monotonic time." )
#pragma message( "Get at least glib 2.53.3 for high precision monotonic time." )
#pragma message( "Using WIN32 API high-resolution performance counter instead." )
	static LARGE_INTEGER Freq_static = {0,0};	/* Performance Frequency */
	LARGE_INTEGER PerfCpt;
	if (! Freq_static.QuadPart)
	{
		QueryPerformanceFrequency(&Freq_static);  /* Frequency value */
	}
	QueryPerformanceCounter(&PerfCpt); /* Counter value */
	return PerfCpt.QuadPart * 1000000.0/Freq_static.QuadPart; /* time in micros */
#else
	return (double) g_get_monotonic_time();
#endif
#else
  /* fallback to less precise clock */
  GTimeVal timeval;

  g_get_current_time (&timeval);

  return (timeval.tv_sec * 1000000.0 + timeval.tv_usec);
#endif
}
```

Reply:
Ok, do it if you want. But in this case I consider the `#pragma messages` to be more confusing to the user than helpful. It just works in the end, no matter how.